packages/contracts-bedrock/contracts/L1/L2OutputOracle.sol

```diff
@@ -131,45 +140,49 @@ contract L2OutputOracle is Initializable, Semver {
     }
 
     /**
-     * @notice Deletes all output proposals after and including the proposal that corresponds to
-     *         the given output index. Only the challenger address can delete outputs.
+     * @notice Deletes an output proposal at the given output l2BlockNumber.
+     *         This function should be called by the dispute game after the game is resolved.
      *
-     * @param _l2OutputIndex Index of the first L2 output to be deleted. All outputs after this
-     *                       output will also be deleted.
+     * @param _l2BlockNumber L2 Block Number whose output to delete.
      */
-    // solhint-disable-next-line ordering
-    function deleteL2Outputs(uint256 _l2OutputIndex) external {
-        require(
-            msg.sender == CHALLENGER,
-            "L2OutputOracle: only the challenger address can delete outputs"
+    function deleteL2Outputs(uint256 _l2BlockNumber) external {
+        // Validate the caller dispute game is complete
+        IDisputeGame caller = IDisputeGame(msg.sender);
+        IDisputeGame game = IDisputeGame(
+            address(
+                DISPUTE_GAME_FACTORY.games(
+                    caller.gameType(),
+                    caller.rootClaim(),
+                    caller.extraData()
+                )
+            )
         );
 
-        // Make sure we're not *increasing* the length of the array.
+        require(msg.sender == address(game), "L2OutputOracle: Unauthorized output deletion.");
+
         require(
-            _l2OutputIndex < l2Outputs.length,
-            "L2OutputOracle: cannot delete outputs after the latest output index"
+            uint8(game.status()) == uint8(GameStatus.CHALLENGER_WINS),
+            "L2OutputOracle: Game incomplete."
         );
 
+        uint256 index = l2OutputIndices[_l2BlockNumber];
+        require(index != 0, "L2OutputOracle: No output exists for the given L2 block number");
+
         // Do not allow deleting any outputs that have already been finalized.
         require(
-            block.timestamp - l2Outputs[_l2OutputIndex].timestamp < FINALIZATION_PERIOD_SECONDS,
+            block.timestamp - l2Outputs[index - 1].timestamp < FINALIZATION_PERIOD_SECONDS,
             "L2OutputOracle: cannot delete outputs that have already been finalized"
         );
 
-        uint256 prevNextL2OutputIndex = nextOutputIndex();
+        // Delete the output root
+        delete l2Outputs[index - 1];
 
-        // Use assembly to delete the array elements because Solidity doesn't allow it.
-        assembly {
-            sstore(l2Outputs.slot, _l2OutputIndex)
-        }
-
-        emit OutputsDeleted(prevNextL2OutputIndex, _l2OutputIndex);
+        uint256 l2OutputsLength = l2Outputs.length;
+        emit OutputsDeleted(l2OutputsLength, l2OutputsLength);
```

Comment:
We should be able to get away with breaking the ABI here to be honest to create an event that makes more sense.

```solidity
bytes32 output = l2Outputs[index - 1];
delete l2Outputs[index - 1];
emit OutputDeleted(output, _l2BlockNumber, msg.sender);
```